Wangscape/utils.h

```diff
@@ -0,0 +1,64 @@
+#pragma once
+#include <type_traits>
+#include <stdexcept>
+
+namespace
+{
+
+template<typename I>
+I _pow(I base, I exp)
+{
+    I result = 1;
+    for (int i = 0; i < std::numeric_limits<I>::digits; i++)
+    {
+        if (exp & 1)
+            result *= base;
+        else if (exp == 0)
+            break;
+        exp >>= 1;
+        base *= base;
+    }
+    return result;
+}
+
+}
+
+template <typename I>
+typename std::enable_if<std::is_unsigned<I>::value, I>::type pow(I base, I exp)
+{
+    if (exp == 0)
+        return 1;
+    else if (base == 0)
+        return 0;
+    else if (base == 1)
+        return 1;
+    else if (exp >= std::numeric_limits<I>::digits)
+        throw std::range_error("Integer pow() with exp >= digits and base > 1");
+    else if (base == 2)
+        return (I)1 << exp;
+    return _pow(base, exp);
+}
+
+template <typename I>
+typename std::enable_if<!std::is_unsigned<I>::value, I>::type pow(I base, I exp)
+{
+    if (exp == 0)
+        return 1;
+    else if (base == 1)
+        return 1;
+    else if (base == -1)
+        return exp % 2 ? -1 : 1;
+    else if (exp < 0)
+        throw std::domain_error("Integer pow() with exp < 0 and |base| != 1");
+    else if (base == 0)
+        return 0;
+    else if (exp > std::numeric_limits<I>::digits)
+        throw std::range_error("Integer pow() with exp > digits and |base| > 1");
+    else if (base == -2)
+        return ((I)1 << exp) * (exp % 2 ? -1 : 1);
```

Comment:
The current usage of `pow` in Wangscape is not in a performance-sensitive area (it's called twice per tileset), but `pow` for integral types is quite a common operation and I expect to reuse this function in future projects. When that happens, I would consider it less of an optimisation issue and more of a bug if I found that obvious O(1) cases were handled using the loop.

Detecting obvious results also allows them to be filtered out when detecting obvious exceptions (`exp >= digits` shouldn't throw when `abs(base) <= 1`, `exp < 0` shouldn't throw when `abs(base) == 1`).

I can imagine a very smart compiler detecting the redundant multiplications when `base` is `-1`, `0`, or `1`, but I wouldn't rely on that behaviour (and I doubt any current compiler would detect when an exponentiation-by-squaring algorithm can be replaced with a bitshift).